server/schedule/merge_checker.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedule
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/pd/server/core"
+	"github.com/pingcap/pd/server/namespace"
+	log "github.com/sirupsen/logrus"
+)
+
+// MergeChecker ensures region to merge with adjacent region when size is small
+type MergeChecker struct {
+	cluster    Cluster
+	classifier namespace.Classifier
+}
+
+// NewMergeChecker creates a merge checker.
+func NewMergeChecker(cluster Cluster, classifier namespace.Classifier) *MergeChecker {
+	return &MergeChecker{
+		cluster:    cluster,
+		classifier: classifier,
+	}
+}
+
+// Check verifies a region's replicas, creating an Operator if need.
+func (m *MergeChecker) Check(region *core.RegionInfo) (*Operator, *Operator) {
+	checkerCounter.WithLabelValues("merge_checker", "check").Inc()
+
+	// region size is not small enough
+	if region.ApproximateSize >= int64(m.cluster.GetMaxMergeRegionSize()) {
+		checkerCounter.WithLabelValues("merge_checker", "small_size").Inc()
+		return nil, nil
+	}
+
+	// skip hot region
+	if m.cluster.IsRegionHot(region.GetId()) {
+		checkerCounter.WithLabelValues("merge_checker", "hot_region").Inc()
+		return nil, nil
+	}
+
+	var target *core.RegionInfo
+	prev, next := m.cluster.GetAdjacentRegions(region)
+
+	peerCount := len(region.Region.GetPeers())
+	// if is not hot region and under same namesapce
+	if prev != nil && !m.cluster.IsRegionHot(prev.GetId()) && m.classifier.AllowMerge(region, prev) {
+		// peer count should equal
+		if peerCount == len(prev.Region.GetPeers()) {
+			target = prev
+		}
+	}
+	if next != nil && !m.cluster.IsRegionHot(next.GetId()) && m.classifier.AllowMerge(region, next) {
```

Comment:
can you extract to the same closure function? I think the previous check can also use the same logic here.